README.md
# Description

[![Latest Version on Packagist](https://img.shields.io/packagist/v/frog/cooperative-bank-sdk.svg?style=flat-square)](https://packagist.org/packages/frog/cooperative-bank-sdk)
[![Build Status](https://img.shields.io/travis/frog/cooperative-bank-sdk/master.svg?style=flat-square)](https://travis-ci.com/github/frogtechnologies/cooperative-bank-sdk)
[![Total Downloads](https://img.shields.io/packagist/dt/frog/cooperative-bank-sdk.svg?style=flat-square)](https://packagist.org/packages/frog/cooperative-bank-sdk)

An SDK to aid interaction with the Cooperative Bank API

## Installation

You can install the package via composer:

```bash
composer require frog/cooperative-bank-sdk
```

### Setup environment variables

```bash
COOP_CONSUMER_KEY=
COOP_CONSUMER_SECRET=
COOP_API_BASE_URL=http://developer.co-opbank.co.ke:8280#Testing
# COOP_API_BASE_URL=https://developer.co-opbank.co.ke:8243 # Production
```

## Usage

```php
use FROG\CooperativeBankSdk\CooperativeBankSdk;
use FROG\CooperativeBankSdk\CoopUtils;

$coop_sdk = new CooperativeBankSDK();

/**
 * Generate an access token
 * NB: Not necessarily used anywhere as the SDK uses it internally to send requests
 *
 */
$access_token = $coop_sdk->generate_access_token();
print_r($access_token);
// {
//   "access_token": "xxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx",
//   "scope": "am_application_scope default",
//   "token_type": "Bearer",
//   "expires_in": 3600
// }

/**
 * Check for the account balance
 *
 * The message reference is optional in the case the client has an internal unique referencing system
 * The SDK implements the php random_bytes function internally to generate a unique number
 * @param string $account_number the cooperative bank account number
 * @param void|string $message_reference an optional message reference
 */
// Option 1: Pass in your unique internal message reference (UUID's don't work though)
$balance = $coop_sdk->check_account_balance(
    "account-number",
    "message-reference",
);

// Or you could use the utility from the package to get a unique message reference before the request is made
$balance = $coop_sdk->check_account_balance(
    "account-number",
    CoopUtils::generate_message_reference(),
);

// Or let the sdk generate one internally and access the message reference from the response. 
// All responses usually have message references
$balance = $coop_sdk->check_account_balance(
    "account-number",
);
print_r($balance);
// {
//   "MessageReference": "MESSAGE_REFERENCE_WILL_APPEAR_HERE",
//   "MessageDateTime": "2021-02-26 15:18:58",
//   "MessageCode": "0",
//   "MessageDescription": "Success",
//   "AccountNumber": "36001873000",
//   "AccountName": "JOE K. DOE",
//   "Currency": "USD",
//   "ProductName": "CURRENT ACCOUNT",
//   "ClearedBalance": "13706.07",
//   "BookedBalance": "75391.31",
//   "BlockedBalance": "27066.64",
//   "AvailableBalance": "21962.96",
//   "ArrearsAmount": "12645.56",
//   "BranchName": "GIGIRI MALL",
//   "BranchSortCode": "11151",
//   "AverageBalance": "27339.95",
//   "UnclearedBalance": "26658.48",
//   "ODLimit": "17614.28",
//   "CreditLimit": "23181.53"
// }

/**
 * Get the full & mini statement of an account
 * @param string $account_number an account number from a cooperative bank branch
 * @param string $start_date the date at which the earliest transaction is to be fetched
 * @param string $end_date the date at which the latest transction is to be fetched
 * @param string $message_reference a unique client generated string to be a reference when a request is sent
 * The message reference is optional in the case the client has an internal unique referencing system
 * The SDK implements the php random_bytes function internally to generate a unique number
*/
$statement = $coop_sdk->get_account_full_statement(
    "account-number",
    "start-date",
    "end-date",
    "message-reference",
);
$statement = $coop_sdk->get_account_mini_statement(
    "account-number",
    "start-date",
    "end-date",
    "message-reference",
);
print_r($statement);
// {
//   "MessageReference": "40ca18c6765086089a1",
//   "MessageDateTime": "2021-02-26 10:27:16",
//   "MessageCode": "0",
//   "MessageDescription": "Success",
//   "AccountNumber": "36001873000",
//   "AccountName": "JOE K. DOE",
//   "Transactions": [
//     {
//       "TransactionID": "50a603a3-2",
//       "TransactionDate": "2021-03-07 23:52:28",
//       "ValueDate": "2021-03-07 23:52:28",
//       "Narration": "Electricity payment",
//       "TransactionType": "D",
//       "ServicePoint": "POS-P0917",
//       "TransactionReference": "8be9b10d-4",
//       "CreditAmount": "0.00",
//       "DebitAmount": "72869.62",
//       "RunningClearedBalance": "22749.45",
//       "RunningBookBalance": "22749.45",
//       "DebitLimit": "0.00",
//       "LimitExpiryDate": "2021-03-07 00:00:00"
//     },
//     ...
//   ]
// }

/**
 * Retrieves the transactions of the specified account
 *
 * @param string $message_reference a unique client generated string to be a reference when a request is sent
 * @param string $account_number an account number from a cooperative bank branch
 * @param string $no_of_transactions a number of transactions to be retrieved (1 - 500)
 */
$transactions = $coop_sdk->get_account_transactions(
    "account-number",
    "no-of-transactions",
    "message-reference",
);
print_r($transactions);
// {
//   "MessageReference": "40ca18c6765086089a1",
//   "MessageDateTime": "2021-02-26 15:18:04",
//   "MessageCode": "0",
//   "MessageDescription": "Success",
//   "AccountNumber": "36001873000",
//   "AccountName": "JOE K. DOE",
//   "Transactions": [
//     {
//       "TransactionID": "fa24e510-c",
//       "TransactionDate": "2021-02-28 20:28:45",
//       "ValueDate": "2021-02-28 20:28:45",
//       "Narration": "Cash Deposit",
//       "TransactionType": "C",
//       "ServicePoint": "AGENT-10001",
//       "TransactionReference": "37e4a38d-9",
//       "CreditAmount": "97705.30",
//       "DebitAmount": "0.00",
//       "RunningClearedBalance": "172774.05",
//       "RunningBookBalance": "172774.05",
//       "DebitLimit": "0.00",
//       "LimitExpiryDate": "2021-02-27 00:00:00"
//     },
//     ...
//   ]
// }

/**
     * Retrieves the status of the specified transaction fetched by a message id
     * @param string $message_reference a unique client generated string to be a reference when a request is sent
     */
$status = $coop_sdk->check_transaction_status("message-reference");
print_r($status);
// {
//     "messageReference": "40ca18c6765086089a1",
//     "messageDateTime": "2021-02-26 15:23:32",
//     "messageCode": "0",
//     "messageDescription": "Full Success",
//     "source": {
//         "accountNumber": "36001873000",
//         "amount": "777",
//         "transactionCurrency": "KES",
//         "narration": "Supplier Payment",
//         "responseCode": "0",
//         "responseDescription": "Success"
//     },
//     "destination": {
//         "referenceNumber": "40ca18c6765086089a1_1",
//         "accountNumber": "54321987654321",
//         "amount": "777",
//         "transactionCurrency": "KES",
//         "narration": "Electricity Payment",
//         "transactionID": "cfea0f9d-c",
//         "responseCode": "0",
//         "responseDescription": "Success"
//     }
// }

/**
 * Retrieves the validity of the specified account
 * 
 * @param string $account_number an account number from a cooperative bank branch
 * @param string $message_reference a unique client generated string to be a reference when a request is sent
 */
$status = $coop_sdk->validate_account("account-number", "message-reference");
print_r($status);
// {
//   "MessageReference": "BTVd6xr7vEX97hWgNqM",
//   "MessageDateTime": "2021-02-26 16:36:39",
//   "MessageCode": "0",
//   "MessageDescription": "VALID ACCOUNT NUMBER"
// }

/**
 * Retrieves the exhange rate for the day for a specific account
 * 
 * @param string $from_currency a valid international currency
 * @param string $to_currency a valid international currency
 * @param string $message_reference a unique client generated string to be a reference when a request is sent
 */
$rate = $coop_sdk->exchange_rate("from-currency", "to-currency", "message-reference");
print_r($rate);
// {
//   "MessageReference": "40ca18c6765086089a1",
//   "MessageDateTime": "2021-02-26 20:31:31",
//   "MessageCode": "0",
//   "MessageDescription": "Success",
//   "FromCurrencyCode": "KES",
//   "ToCurrencyCode": "USD",
//   "RateType": "SPOT",
//   "Rate": "104.35",
//   "Tolerance": "6",
//   "MultiplyDivide": "D"
// }
```

### Testing

`NB`: The primary development machine is windows, ensure to change the path in the composer.json script section to conform to the operating system

Tests are powered by [PestPHP](https://pestphp.com/)

```bash
composer test
```

### Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information what has changed recently.

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

### Security

If you discover any security related issues, please email milleradulu@gmail.com instead of using the issue tracker.

## Credits

-   [Miller Adulu](https://github.com/milleradulu)
-   [All Contributors](../../contributors)

## License

The MIT. Please see [License File](LICENSE.md) for more information.

## PHP Package Boilerplate

This package was generated using the [PHP Package Boilerplate](https://laravelpackageboilerplate.com).


CHANGELOG.md
# Changelog

All notable changes to `cooperative-bank-sdk` will be documented in this file

## 1.0.0 - 201X-XX-XX

- initial release


tests/Pest.php
<?php

/*
|--------------------------------------------------------------------------
| Test Case
|--------------------------------------------------------------------------
|
| The closure you provide to your test functions is always bound to a specific PHPUnit test
| case class. By default, that class is "PHPUnit\Framework\TestCase". Of course, you may
| need to change it using the "uses()" function to bind a different classes or traits.
|
*/

// uses(Tests\TestCase::class)->in('Feature');

/*
|--------------------------------------------------------------------------
| Expectations
|--------------------------------------------------------------------------
|
| When you're writing tests, you often need to check that values meet certain conditions. The
| "expect()" function gives you access to a set of "expectations" methods that you can use
| to assert different things. Of course, you may extend the Expectation API at any time.
|
*/



/*
|--------------------------------------------------------------------------
| Functions
|--------------------------------------------------------------------------
|
| While Pest is very powerful out-of-the-box, you may have some testing code specific to your
| project that you don't want to repeat in every file. Here you can also expose helpers as
| global functions to help you to reduce the number of lines of code in your test files.
|
*/

use FROG\CooperativeBankSdk\CooperativeBankEndpoint;

/**
 * @return array<int, mixed> 
 */
function get_valid_auth_options(): array
{
    $consumer_key = CooperativeBankEndpoint::DEFAULT_CONSUMER_KEY;
    $consumer_secret = CooperativeBankEndpoint::DEFAULT_CONSUMER_SECRET;
    $coop_base_url = CooperativeBankEndpoint::DEFAULT_BASE_URL;

    $base_64_auth = base64_encode("$consumer_key:$consumer_secret");

    $auth_headers = ["Authorization: Basic {$base_64_auth}"];
    $auth_data = "grant_type=client_credentials";

    return [
        CURLOPT_URL => $coop_base_url . '/token',
        CURLOPT_HTTPHEADER => $auth_headers,
        CURLOPT_SSL_VERIFYPEER  => false,
        CURLOPT_RETURNTRANSFER => true,
        CURLOPT_POST => true,
        CURLOPT_POSTFIELDS => $auth_data,
    ];
}

/**
 * @param string $access_token
 * @param string $path
 * @param array<string, mixed> $body
 * @return array<int, mixed> 
 */
function get_valid_req_options(
    string $access_token,
    string $path,
    array $body
): array {
    $auth_headers = [
        "Authorization: Bearer {$access_token}",
        "Content-Type: application/json",
    ];

    return [
        CURLOPT_URL => CooperativeBankEndpoint::DEFAULT_BASE_URL . $path,
        CURLOPT_HTTPHEADER => $auth_headers,
        CURLOPT_SSL_VERIFYPEER  => false,
        CURLOPT_RETURNTRANSFER => true,
        CURLOPT_POST => true,
        CURLOPT_POSTFIELDS => json_encode($body),
    ];
}


function confirm_standard_response(object $result): void
{
    expect($result)->toBeObject();
    expect($result)->toHaveProperty('MessageReference');
    expect($result)->toHaveProperty('MessageDateTime');
    expect($result)->toHaveProperty('MessageCode');
    expect($result)->toHaveProperty('MessageDescription');
}


function confirm_standard_response_v2(object $result): void
{
    expect($result)->toBeObject();
    expect($result)->toHaveProperty('messageReference');
    expect($result)->toHaveProperty('messageDateTime');
    expect($result)->toHaveProperty('messageCode');
    expect($result)->toHaveProperty('messageDescription');
}


tests/ExampleTest.php
<?php

namespace MillerAdulu\CooperativeBankSdk\Tests;

use PHPUnit\Framework\TestCase;

class ExampleTest extends TestCase
{
    /** @test */
    public function true_is_true()
    {
        $this->assertTrue(true);
    }
}


tests/Unit/ApiAuthTest.php
<?php

use FROG\CooperativeBankSdk\CooperativeBankSdk;
use FROG\PhpCurlSAI\SAI_CurlStub;

it('can generate an access token', function () {

    $cURL = new SAI_CurlStub();
    $sdk = new CooperativeBankSdk($cURL);
    $cURL->setResponse(
        '{"access_token":"4fbb7b9e-9c73-3155-ac14-c14fe744d104","scope":"am_application_scope default","token_type":"Bearer","expires_in":3600}',
        get_valid_auth_options()
    );


    $result = $sdk->generate_access_token();

    expect($result)->toBeObject();
    expect($result)->toHaveProperty('access_token');
    expect($result)->toHaveProperty('scope');
    expect($result)->toHaveProperty('token_type');
    expect($result)->toHaveProperty('expires_in');
});


src/CooperativeBankEndpoint.php
<?php

namespace FROG\CooperativeBankSdk;

/**
 * Specifies the default variables and endpoint to use in the testing environment.
 * 
 */
class CooperativeBankEndpoint
{
    const DEFAULT_BASE_URL = "http://developer.co-opbank.co.ke:8280";
    const DEFAULT_CONSUMER_KEY = "zuP_MW9YUs69mpXPZaubHnEo1x8a";
    const DEFAULT_CONSUMER_SECRET = "lWzT7h9UGmsflIP0xzjCQSoV77wa";

    const ACCESS_TOKEN = "/token";
    const ACCOUNT_BALANCE = "/Enquiry/AccountBalance/1.0.0";

    // NB: As of 26th February 2021, these work in the inverse manner
    const FULL_STATEMENT = "/Enquiry/MiniStatement/Account/1.0.0";
    const MINI_STATEMENT = "/Enquiry/FullStatement/Account/1.0.0";

    const ACCOUNT_TRANSACTIONS = "/Enquiry/AccountTransactions/1.0.0";
    const TRANSACTION_STATUS = "/Enquiry/TransactionStatus/2.0.0";
    const VALIDATE_ACCOUNT = "/Enquiry/Validation/Account/1.0.0";

    const EXCHANGE_RATE = "/Enquiry/ExchangeRate/1.0.0";
}


src/CooperativeBankSdk.php
<?php

namespace FROG\CooperativeBankSdk;

use Dotenv\Dotenv;
use FROG\PhpCurlSAI\SAI_Curl;
use FROG\PhpCurlSAI\SAI_CurlInterface;

/**
 * Contains the methods to call so that the interaction with the API can happen properly
 */
class  CooperativeBankSdk
{
    protected SAI_CurlInterface $cURL;

    // Hard coded defaults for the testing environment
    protected string $default_base_url = CooperativeBankEndpoint::DEFAULT_BASE_URL;

    /**
     * Sets up the class to use a stubbed cURL implementation class
     * 
     * void|SAI_CurlInterface $cURl: A curl implementation of the SAI_CurlInterface
     */
    public function __construct(
        SAI_CurlInterface $cURL = null
    ) {
        if ($cURL == null) $this->cURL = new SAI_Curl();
        else
            $this->cURL = $cURL;
    }

    /**
     * Generates an access token from the Cooperative Bank API
     * 
     * @return object an stdClass of the response from the API 
     * {
     * "access_token": "xxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx",
     * "scope": "am_application_scope default",
     * "token_type": "Bearer",
     * "expires_in": 3600
     * }
     */
    public function generate_access_token(): ?object
    {
        try {
            $dotenv = Dotenv::createImmutable(__DIR__);
            $dotenv->load();
        } catch (\Throwable $th) {
            //throw $th;
        }

        $consumer_key = $_ENV['COOP_CONSUMER_KEY'] ?? CooperativeBankEndpoint::DEFAULT_CONSUMER_KEY;
        $consumer_secret = $_ENV['COOP_CONSUMER_SECRET'] ?? CooperativeBankEndpoint::DEFAULT_CONSUMER_SECRET;
        $coop_base_url = $_ENV['COOP_API_BASE_URL'] ?? $this->default_base_url;

        $base_64_auth = base64_encode("$consumer_key:$consumer_secret");

        $auth_headers = ["Authorization: Basic {$base_64_auth}"];
        $auth_data = "grant_type=client_credentials";

        $options = [
            CURLOPT_URL => $coop_base_url . CooperativeBankEndpoint::ACCESS_TOKEN,
            CURLOPT_HTTPHEADER => $auth_headers,
            CURLOPT_SSL_VERIFYPEER  => false,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_POST => true,
            CURLOPT_POSTFIELDS => $auth_data,
        ];

        $ch = $this->cURL->curl_init();
        $this->cURL->curl_setopt_array($ch, $options);
        $response = $this->cURL->curl_exec($ch);

        if ($response === false) {
            $result = $this->cURL->curl_error($ch);
        } else {
            $result = json_decode($response);
        }

        // print_r($response);

        $this->cURL->curl_close($ch);

        return $result;
    }

    /**
     * Retrieves the balance of the specified account
     * 
     * @param string $account_number an account number from a cooperative bank branch
     * @param string $message_reference a unique client generated string to be a reference when a request is sent
     * @return object an stdClass of the response from the API
     */
    public function check_account_balance(
        string $account_number,
        ?string $message_reference = null
    ): ?object {
        try {
            $dotenv = Dotenv::createImmutable(__DIR__);
            $dotenv->load();
        } catch (\Throwable $th) {
            //throw $th;
        }

        $_message_reference = $message_reference ?? CoopUtils::generate_message_reference();

        $_coop_base_url = $_ENV['COOP_API_BASE_URL'] ?? $this->default_base_url;
        $_token_result = $this->generate_access_token();

        $auth_headers = [
            "Authorization: Bearer {$_token_result->access_token}",
            "Content-Type: application/json",
        ];

        $request_body = [
            'MessageReference' => $_message_reference,
            'AccountNumber' => $account_number,
        ];

        $options = [
            CURLOPT_URL => $_coop_base_url . CooperativeBankEndpoint::ACCOUNT_BALANCE,
            CURLOPT_HTTPHEADER => $auth_headers,
            CURLOPT_SSL_VERIFYPEER  => false,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_POST => true,
            CURLOPT_POSTFIELDS => json_encode($request_body),
        ];

        $ch = $this->cURL->curl_init();
        $this->cURL->curl_setopt_array($ch, $options);
        $response = $this->cURL->curl_exec($ch);

        if ($response === false) {
            $result = $this->cURL->curl_error($ch);
        } else {
            $result = json_decode($response);
        }


        $this->cURL->curl_close($ch);


        return $result;
    }

    /**
     * Retrieves the full statement of the specified account
     * 
     * @param string $account_number an account number from a cooperative bank branch
     * @param string $start_date the date at which the earliest transaction is to be fetched
     * @param string $end_date the date at which the latest transction is to be fetched
     * @param string $message_reference a unique client generated string to be a reference when a request is sent
     * @return object an stdClass of the response from the API
     */
    public function get_account_full_statement(
        string $account_number,
        string $start_date,
        string $end_date,
        ?string $message_reference = null
    ): ?object {

        try {
            $dotenv = Dotenv::createImmutable(__DIR__);
            $dotenv->load();
        } catch (\Throwable $th) {
            //throw $th;
        }

        $_coop_base_url = $_ENV['COOP_API_BASE_URL'] ?? $this->default_base_url;
        $_token_result = $this->generate_access_token();
        $_message_reference = $message_reference ?? CoopUtils::generate_message_reference();

        $auth_headers = [
            "Authorization: Bearer {$_token_result->access_token}",
            "Content-Type: application/json",
        ];

        $request_body = [
            'MessageReference' => $_message_reference,
            'AccountNumber' => $account_number,
            "StartDate" => $start_date,
            "EndDate" => $end_date,
        ];

        $options = [
            CURLOPT_URL => $_coop_base_url . CooperativeBankEndpoint::FULL_STATEMENT,
            CURLOPT_HTTPHEADER => $auth_headers,
            CURLOPT_SSL_VERIFYPEER  => false,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_POST => true,
            CURLOPT_POSTFIELDS => json_encode($request_body),
        ];

        $ch = $this->cURL->curl_init();
        $this->cURL->curl_setopt_array($ch, $options);
        $response = $this->cURL->curl_exec($ch);

        if ($response === false) {
            $result = $this->cURL->curl_error($ch);
        } else {
            $result = json_decode($response);
        }

        $this->cURL->curl_close($ch);

        return $result;
    }

    /**
     * Retrieves the mini statement of the specified account
     * 
     * @param string $account_number an account number from a cooperative bank branch
     * @param string $start_date the date at which the earliest transaction is to be fetched
     * @param string $end_date the date at which the latest transaction is to be fetched
     * @param string $message_reference a unique client generated string to be a reference when a request is sent
     * @return object an stdClass of the response from the API
     */
    public function get_account_mini_statement(
        string $account_number,
        string $start_date,
        string $end_date,
        ?string $message_reference = null
    ): ?object {

        try {
            $dotenv = Dotenv::createImmutable(__DIR__);
            $dotenv->load();
        } catch (\Throwable $th) {
            //throw $th;
        }

        $_coop_base_url = $_ENV['COOP_API_BASE_URL'] ?? $this->default_base_url;
        $_token_result = $this->generate_access_token();
        $_message_reference = $message_reference ?? CoopUtils::generate_message_reference();

        $auth_headers = [
            "Authorization: Bearer {$_token_result->access_token}",
            "Content-Type: application/json",
        ];

        $request_body = [
            'MessageReference' => $_message_reference,
            'AccountNumber' => $account_number,
            "StartDate" => $start_date,
            "EndDate" => $end_date,
        ];

        $options = [
            CURLOPT_URL => $_coop_base_url . CooperativeBankEndpoint::MINI_STATEMENT,
            CURLOPT_HTTPHEADER => $auth_headers,
            CURLOPT_SSL_VERIFYPEER  => false,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_POST => true,
            CURLOPT_POSTFIELDS => json_encode($request_body),
        ];

        $ch = $this->cURL->curl_init();
        $this->cURL->curl_setopt_array($ch, $options);
        $response = $this->cURL->curl_exec($ch);

        if ($response === false) {
            $result = $this->cURL->curl_error($ch);
        } else {
            $result = json_decode($response);
        }

        $this->cURL->curl_close($ch);

        return $result;
    }

    /**
     * Retrieves the transactions of the specified account
     * 
     * @param string $message_reference a unique client generated string to be a reference when a request is sent
     * @param string $account_number an account number from a cooperative bank branch
     * @param string $no_of_transactions a number of transactions to be retrieved (1 - 500)
     * @return object an stdClass of the response from the API
     */
    public function get_account_transactions(
        string $account_number,
        string $no_of_transactions,
        ?string $message_reference = null
    ): ?object {
        try {
            $dotenv = Dotenv::createImmutable(__DIR__);
            $dotenv->load();
        } catch (\Throwable $th) {
            //throw $th;
        }

        $_coop_base_url = $_ENV['COOP_API_BASE_URL'] ?? $this->default_base_url;
        $_token_result = $this->generate_access_token();
        $_message_reference = $message_reference ?? CoopUtils::generate_message_reference();

        $auth_headers = [
            "Authorization: Bearer {$_token_result->access_token}",
            "Content-Type: application/json",
        ];

        $request_body = [
            'MessageReference' => $_message_reference,
            'AccountNumber' => $account_number,
            "NoOfTransactions" => $no_of_transactions,
        ];

        $options = [
            CURLOPT_URL => $_coop_base_url . CooperativeBankEndpoint::ACCOUNT_TRANSACTIONS,
            CURLOPT_HTTPHEADER => $auth_headers,
            CURLOPT_SSL_VERIFYPEER  => false,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_POST => true,
            CURLOPT_POSTFIELDS => json_encode($request_body),
        ];

        $ch = $this->cURL->curl_init();
        $this->cURL->curl_setopt_array($ch, $options);
        $response = $this->cURL->curl_exec($ch);

        if ($response === false) {
            $result = $this->cURL->curl_error($ch);
        } else {
            $result = json_decode($response);
        }

        $this->cURL->curl_close($ch);

        return $result;
    }

    /**
     * Retrieves the status of the specified transaction fetched by a message id
     * 
     * @param string $message_reference a unique client generated string to be a reference when a request is sent
     * @return object an stdClass of the response from the API
     */
    public function check_transaction_status(
        string $message_reference
    ): ?object {
        try {
            $dotenv = Dotenv::createImmutable(__DIR__);
            $dotenv->load();
        } catch (\Throwable $th) {
            //throw $th;
        }

        $coop_base_url = $_ENV['COOP_API_BASE_URL'] ?? $this->default_base_url;
        $token_result = $this->generate_access_token();

        $auth_headers = [
            "Authorization: Bearer {$token_result->access_token}",
            "Content-Type: application/json",
        ];

        $request_body = [
            'MessageReference' => $message_reference,
        ];

        $options = [
            CURLOPT_URL => $coop_base_url . CooperativeBankEndpoint::TRANSACTION_STATUS,
            CURLOPT_HTTPHEADER => $auth_headers,
            CURLOPT_SSL_VERIFYPEER  => false,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_POST => true,
            CURLOPT_POSTFIELDS => json_encode($request_body),
        ];

        $ch = $this->cURL->curl_init();
        $this->cURL->curl_setopt_array($ch, $options);
        $response = $this->cURL->curl_exec($ch);

        if ($response === false) {
            $result = $this->cURL->curl_error($ch);
        } else {
            $result = json_decode($response);
        }

        $this->cURL->curl_close($ch);

        return $result;
    }

    /**
     * Retrieves the validity of the specified account
     * 
     * @param string $account_number an account number from a cooperative bank branch
     * @param string $message_reference a unique client generated string to be a reference when a request is sent
     * @return object an stdClass of the response from the API
     */
    public function validate_account(
        string $account_number,
        ?string $message_reference = null
    ): ?object {
        try {
            $dotenv = Dotenv::createImmutable(__DIR__);
            $dotenv->load();
        } catch (\Throwable $th) {
            //throw $th;
        }

        $_coop_base_url = $_ENV['COOP_API_BASE_URL'] ?? $this->default_base_url;
        $_token_result = $this->generate_access_token();
        $_message_reference = $message_reference ?? CoopUtils::generate_message_reference();

        $auth_headers = [
            "Authorization: Bearer {$_token_result->access_token}",
            "Content-Type: application/json",
        ];

        $request_body = [
            'MessageReference' => $_message_reference,
            'AccountNumber' => $account_number,
        ];

        $options = [
            CURLOPT_URL => $_coop_base_url . CooperativeBankEndpoint::VALIDATE_ACCOUNT,
            CURLOPT_HTTPHEADER => $auth_headers,
            CURLOPT_SSL_VERIFYPEER  => false,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_POST => true,
            CURLOPT_POSTFIELDS => json_encode($request_body),
        ];

        $ch = $this->cURL->curl_init();
        $this->cURL->curl_setopt_array($ch, $options);
        $response = $this->cURL->curl_exec($ch);

        if ($response === false) {
            $result = $this->cURL->curl_error($ch);
        } else {
            $result = json_decode($response);
        }


        $this->cURL->curl_close($ch);

        return $result;
    }


    /**
     * Retrieves the exhange rate for the day for a specific account
     * 
     * @param string $from_currency a valid international currency
     * @param string $to_currency a valid international currency
     * @param string $message_reference a unique client generated string to be a reference when a request is sent
     * @return object an stdClass of the response from the API
     */
    public function exchange_rate(
        string $from_currency,
        string $to_currency,
        ?string $message_reference = null
    ) {

        try {
            $dotenv = Dotenv::createImmutable(__DIR__);
            $dotenv->load();
        } catch (\Throwable $th) {
            //throw $th;
        }

        $_coop_base_url = $_ENV['COOP_API_BASE_URL'] ?? $this->default_base_url;
        $_token_result = $this->generate_access_token();
        $_message_reference = $message_reference ?? CoopUtils::generate_message_reference();

        $auth_headers = [
            "Authorization: Bearer {$_token_result->access_token}",
            "Content-Type: application/json",
        ];

        $request_body = [
            'MessageReference' => $_message_reference,
            'FromCurrencyCode' => $from_currency,
            'ToCurrencyCode' => $to_currency,
        ];

        $options = [
            CURLOPT_URL => $_coop_base_url . CooperativeBankEndpoint::EXCHANGE_RATE,
            CURLOPT_HTTPHEADER => $auth_headers,
            CURLOPT_SSL_VERIFYPEER  => false,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_POST => true,
            CURLOPT_POSTFIELDS => json_encode($request_body),
        ];

        $ch = $this->cURL->curl_init();
        $this->cURL->curl_setopt_array($ch, $options);
        $response = $this->cURL->curl_exec($ch);

        if ($response === false) {
            $result = $this->cURL->curl_error($ch);
        } else {
            $result = json_decode($response);
        }

        $this->cURL->curl_close($ch);

        return $result;
    }
}
